### PR TITLE
Swift - UniFFI boundary - update `into_custom` and `from_custom` especially for date.

### PR DESCRIPTION
**Pull request by CyonAlexRDX**
Update to use ISO8601DateFormatter.

I've tried running the "switch to 12 hour format"-test on device and with these changes I was unable to reproduce the crash, so these changes seems to fix the bug.

### CODE REVIEW

File: apple/Tests/TestCases/Profile/DeviceInfoTests.swift [hunk 1]
```diff
@@ -76,6 +76,50 @@ final class DeviceInfoTests: Test<DeviceInfo> {
...
+	func test_date() {
+		
+		/// This matches `[bindings.swift.custom_types.Timestamp]`
+		/// inside `uniffi.toml` with the only difference that we use `$0` here but
+		/// the toml uses `{}`;
+		let into_custom: (String) -> Date = {let s = $0;let df = ISO8601DateFormatter();let fo = ISO8601DateFormatter.Options.withFractionalSeconds;df.formatOptions.insert(fo);func t() -> Date? {df.date(from: s)};if let d = t() {return d};df.formatOptions.remove(fo);return t()!}
+		
+		/// This matches `[bindings.swift.custom_types.Timestamp]`
+		/// inside `uniffi.toml`
+		let from_custom: (Date) -> String = { let df = ISO8601DateFormatter();df.formatOptions.insert(.withFractionalSeconds);return df.string(from: $0) }
```
Comment (kugel3) [inline apple/Tests/TestCases/Profile/DeviceInfoTests.swift:89]: Could just be `$0.formatted(.iso8601.time(includingFractionalSeconds: true))`

File: uniffi.toml [hunk 2]
```diff
@@ -44,8 +44,26 @@ from_custom = "{}.toString()"
...
 imports = ["Foundation"]
-into_custom = "{let df = DateFormatter();df.dateFormat = \"yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ\";return df.date(from: {})!}()"
-from_custom = "{let df = DateFormatter();df.dateFormat = \"yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ\";return df.string(from: {})}()"
+into_custom = """
+{
+    let s = {}
+    let df = ISO8601DateFormatter()
+    let fo = ISO8601DateFormatter.Options.withFractionalSeconds
+    df.formatOptions.insert(fo)
+    func t() -> Date? {df.date(from: s)}
```
Comment (kugel3) [inline uniffi.toml:53]: Not sure it improves readability to define this function, could just call `df.date(from: s)` directly?

File: apple/Tests/TestCases/Profile/DeviceInfoTests.swift [hunk 3]
```diff
@@ -76,6 +76,50 @@ final class DeviceInfoTests: Test<DeviceInfo> {
...
+			XCTAssertEqual(rhsString, rhs)
+			
+			XCTAssertEqual(
+				into_custom(lhsString),
+				into_custom(rhsString)
+			)
+		}
+		
+		let vectors = [
+			("2023-12-24T17:13:56.123456Z", "2023-12-24T17:13:56.123Z"), // precision lost (which is OK)
```
Comment (kugel3) [inline apple/Tests/TestCases/Profile/DeviceInfoTests.swift:116]: Could add one where it's rounded up, to make sure it doesn't just truncate

File: uniffi.toml (hunk 2 above)
Comment (matiasbzurovski) [reply uniffi.toml:53]: agree

File: apple/Tests/TestCases/Profile/DeviceInfoTests.swift (hunk 1 above)
Comment (CyonAlexRDX) [reply apple/Tests/TestCases/Profile/DeviceInfoTests.swift:89]: nope, this produces *only* time - `17:13:56.123`.

File: apple/Tests/TestCases/Profile/DeviceInfoTests.swift (hunk 3 above)
Comment (CyonAlexRDX) [reply apple/Tests/TestCases/Profile/DeviceInfoTests.swift:116]: but it *does* truncate, this "vector" fails:

```swift
     ("2023-12-24T17:13:56.123999Z", "2023-12-24T17:13:56.124Z"),
```

but it is ok, we dont need that precision :D

File: apple/Tests/TestCases/Profile/DeviceInfoTests.swift [hunk 4]
```diff
@@ -76,6 +76,50 @@ final class DeviceInfoTests: Test<DeviceInfo> {
         }
         """, expected: nil)
 	}
+	
+	func test_date() {
```
Comment (matiasbzurovski) [inline apple/Tests/TestCases/Profile/DeviceInfoTests.swift:80]: Why is this under `DeviceInfoTests`?

**Review by matiasbzurovski — APPROVED** [review]

File: apple/Tests/TestCases/Profile/DeviceInfoTests.swift (hunk 3 above)
Comment (kugel3) [reply apple/Tests/TestCases/Profile/DeviceInfoTests.swift:116]: OK, but then include one that tests that it truncates I guess.

Comment (CyonAlexRDX) [reply apple/Tests/TestCases/Profile/DeviceInfoTests.swift:116]: the one above performs truncation..?

```
("2023-12-24T17:13:56.123456Z", "2023-12-24T17:13:56.123Z"),
```

please suggest an exact vector you wanna use if not that? :)

**Pull request merged**